core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/abstractsyntaxtree/constraintcomponents/AndConstraintComponent.java

```diff
@@ -0,0 +1,128 @@
+package org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.constraintcomponents;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.eclipse.rdf4j.model.IRI;
+import org.eclipse.rdf4j.model.Model;
+import org.eclipse.rdf4j.model.Resource;
+import org.eclipse.rdf4j.model.vocabulary.SHACL;
+import org.eclipse.rdf4j.repository.RepositoryConnection;
+import org.eclipse.rdf4j.sail.shacl.AST.ShaclProperties;
+import org.eclipse.rdf4j.sail.shacl.ConnectionsGroup;
+import org.eclipse.rdf4j.sail.shacl.ShaclSail;
+import org.eclipse.rdf4j.sail.shacl.SourceConstraintComponent;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.Cache;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.HelperTool;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.NodeShape;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.PropertyShape;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.Shape;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.planNodes.DebugPlanNode;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.planNodes.EmptyNode;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.planNodes.PlanNode;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.planNodes.PlanNodeProvider;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.planNodes.UnionNode;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.planNodes.Unique;
+import org.eclipse.rdf4j.sail.shacl.abstractsyntaxtree.targets.TargetChain;
+
+public class AndConstraintComponent extends AbstractConstraintComponent {
+	List<Shape> and;
+
+	public AndConstraintComponent(Resource id, RepositoryConnection connection,
+			Cache cache, ShaclSail shaclSail) {
```

Comment:
Ah, this suggests that this AST is still tightly coupled to the ShaclSail, in which case separating it out to a generic package probably makes little sense.

Reply:
It's not that tightly coupled. The SHACL sail is only used to get hold of some user settings. Could make an interface for that instead.